feat(datasource/graphene) find path tool

chrisj opened the pull request:
This tool finds a path between 2 points on the same segment.

I crudely copied AnnotationLayerView.makeAnnotationListElement to create some annotation widgets for this tool. It seems to work fairly well as a UI element. Do you think I should continue and fully move makeAnnotationListElement outside of AnnotationLayerView? Or would it be better to create my own element.

<img width="558" alt="Screen Shot 2023-07-19 at 10 15 42 AM" src="https://github.com/google/neuroglancer/assets/188155/c02c8d80-a124-4cf8-a985-680a9166c044">

chrisj committed:
feat(datasource/graphene): add graphene find path tool

jbms commented:
Yes, factoring out makeAnnotationListElement seems to make sense to me.

chrisj commented:
@jbms this should be ready now

chrisj commented on src/neuroglancer/ui/annotations.ts:
```diff
@@ -461,7 +461,25 @@ export class AnnotationLayerView extends Tab {
+        selectionState.annotationId === annotation.id) {
+      element.classList.add('neuroglancer-annotation-selected');
+    }
+    return element;
```
moved a few things outside of `makeAnnotationListElement` that were more closely related to the list

jbms commented on src/neuroglancer/datasource/graphene/frontend.ts:
```diff
@@ -61,6 +65,7 @@ import {Uint64} from 'neuroglancer/util/uint64';
 import {makeDeleteButton} from 'neuroglancer/widget/delete_button';
 import {DependentViewContext} from 'neuroglancer/widget/dependent_view_widget';
 import {makeIcon} from 'neuroglancer/widget/icon';
+import {CancellationToken, CancellationTokenSource} from 'src/neuroglancer/util/cancellation';
```
fix src/neuroglancer/ path to neuroglancer/

chrisj replied:
Done, I'm also going to put up a pull request to fix some other imports.

chrisj commented:
This PR is ready now

chrisj committed:
fix(annotations): use subsubsource id when determining the selected annotation layer state

chrisj commented:
I moved the setting of hoverState back into makeAnnotationListElement, this allows our use of the annotation list item in the find path tool (which is now treated more like a widget) to highlight the corresponding row in the layer annotation list.

I also added a fix for the selected color not appearing when you have multiple annotation states with the same subsource id.

chrisj committed:
fix: find path annotation widget now uses max column width
switched event listeners in find path tool to use activation.registerDisposer
setting annotationSubsubsourceId for captureSelectionState

chrisj committed:
fix: find path tool annotation widget now causes a hover effect in the annotation list to match the behavior of the segment widget

chrisj commented:
@seankmartin if you have time to review this, it would be great to finally close this
the only state changes are specific to graphene with the addition of a find path state

The major change outside of graphene here is migrating `makeAnnotationListElement` to be more of a widget like the segment widget.

seankmartin commented on src/datasource/graphene/frontend.ts:
```diff
@@ -958,6 +1023,146 @@ class MergeState extends RefCounted implements Trackable {
+    this.registerDisposer(this.centroids.changed.add(this.changed.dispatch));
+  }
+
+  get path() {
```
I experimented with changing the series of lines to a polyline to stress test the polyline implementation. The version here was done pretty quickly so might need some slight changes, but I think it's a bit cleaner. Not sure if you've something downstream that uses the lines, but if not, it might be nice to change to using polylines

```
diff --git a/src/datasource/graphene/frontend.ts b/src/datasource/graphene/frontend.ts
index 6c892e99..46c03c7a 100644
--- a/src/datasource/graphene/frontend.ts
+++ b/src/datasource/graphene/frontend.ts
@@ -27,6 +27,7 @@ import type {
   AnnotationSource,
   Line,
   Point,
+  PolyLine,
 } from "#src/annotation/index.js";
 import {
   AnnotationType,
@@ -1055,43 +1056,40 @@ class FindPathState extends RefCounted implements Trackable {
   }
 
   get path() {
-    const path: Line[] = [];
     const {
       source: { value: source },
       target: { value: target },
       centroids: { value: centroids },
     } = this;
+    const path: PolyLine = {
+      points: [],
+      type: AnnotationType.POLYLINE,
+      id: "",
+      properties: [],
+    };
     if (!source || !target || centroids.length === 0) {
       return path;
     }
-    for (let i = 0; i < centroids.length - 1; i++) {
-      const pointA = centroids[i];
-      const pointB = centroids[i + 1];
-      const line: Line = {
-        pointA: vec3.fromValues(pointA[0], pointA[1], pointA[2]),
-        pointB: vec3.fromValues(pointB[0], pointB[1], pointB[2]),
-        id: "",
-        type: AnnotationType.LINE,
-        properties: [],
-      };
-      path.push(line);
+    path.points.push(
+      new Float32Array([
+        source.position[0],
+        source.position[1],
+        source.position[2],
+      ]),
+    );
+    for (const centroid of centroids) {
+      path.points.push(
+        new Float32Array([centroid[0], centroid[1], centroid[2]]),
+      );
     }
-    const firstLine: Line = {
-      pointA: source.position,
-      pointB: path[0].pointA,
-      id: "",
-      type: AnnotationType.LINE,
-      properties: [],
-    };
-    const lastLine: Line = {
-      pointA: path[path.length - 1].pointB,
-      pointB: target.position,
-      id: "",
-      type: AnnotationType.LINE,
-      properties: [],
-    };
-
-    return [firstLine, ...path, lastLine];
+    path.points.push(
+      new Float32Array([
+        target.position[0],
+        target.position[1],
+        target.position[2],
+      ]),
+    );
+    return path;
   }
 
   replaceSegments(oldValues: Uint64Set, newValues: Uint64Set) {
@@ -1475,8 +1473,8 @@ class GraphConnection extends SegmentationGraphSourceConnection {
           annotationSource.delete(annotationSource.getReference(annotation.id));
         }
       }
-      for (const line of path) {
-        annotationSource.add(line);
+      if (path.points.length > 0) {
+        annotationSource.add(path);
       }
     };
     this.registerDisposer(findPathState.changed.add(findPathChanged));
```

chrisj replied:
@seankmartin thanks for reviewing and great idea, converted it to use polyline

seankmartin reviewed (commented):
Refactoring of the annotation code looks fine to me and so does the feature. One comment I have is about maybe moving from Line to Polyline but outside of that looks good to go

chrisj committed:
feat(datasource/graphene): use polyline for find path tool
